gnc/sub8_perception/include/sub8_pcl/cv_tools.hpp

```diff
@@ -9,4 +19,49 @@ typedef std::vector<cv::Point> Contour;
 
 // Compute the centroid of an OpenCV contour (Not templated)
 cv::Point contour_centroid(Contour& contour);
-}
\ No newline at end of file
+
+struct ImageWithCameraInfo
+{	
+	/**
+		Packages corresponding  sensor_msgs::ImageConstPtr and sensor_msgs::CameraInfoConstPtr info_msg
+		into one object. Containers of these objects can be sorted by their image_time attribute
+	*/
+	public:
+		ImageWithCameraInfo(sensor_msgs::ImageConstPtr image_msg, sensor_msgs::CameraInfoConstPtr info_msg);
+		sensor_msgs::ImageConstPtr image_msg;
+		sensor_msgs::CameraInfoConstPtr info_msg;
+		ros::Time image_time;
+		bool operator <(const ImageWithCameraInfo& right) const {
+			return this->image_time < right.image_time;
+		}
+};
+
+
+class FrameHistory
+{
+	/**
+		Object that subscribes itself to an image topic and stores up to a user defined
+		number of ImageWithCameraInfo objects. The frame history can then be retrieved 
+		in whole or just a portion.
+	*/
+public:
+
+	FrameHistory(std::string img_topic, unsigned int hist_size);
+	~FrameHistory();
+	void image_callback(const sensor_msgs::ImageConstPtr &image_msg, const sensor_msgs::CameraInfoConstPtr &info_msg);
+	std::vector<ImageWithCameraInfo> get_frame_history(unsigned int frames_requested);
+	int frames_available();
+
+	const std::string topic_name;
+	const unsigned int history_size;
+
+private:
+	ros::NodeHandle nh;
+	image_transport::CameraSubscriber _image_sub;
+	image_transport::ImageTransport _image_transport;
+	std::vector<ImageWithCameraInfo> _frame_history_ring_buffer;
```

Comment:
Probably want to implement a ring buffer with a [deque](http://en.cppreference.com/w/cpp/container/deque), which is like a vector optimized for right-append and left-pop